extensions/ql-vscode/src/astViewer.ts

```diff
@@ -83,33 +96,83 @@ class AstViewerDataProvider implements vscode.TreeDataProvider<AstItem | RootAst
 }
 
 export class AstViewer {
-  private treeView: vscode.TreeView<AstItem | RootAstItem>;
+  private treeView: TreeView<AstItem | RootAstItem>;
   private treeDataProvider: AstViewerDataProvider;
+  private currentFile: string | undefined;
 
-  constructor() {
+  constructor(ctx: ExtensionContext) {
     this.treeDataProvider = new AstViewerDataProvider();
-    this.treeView = vscode.window.createTreeView('codeQLAstViewer', {
+    this.treeView = window.createTreeView('codeQLAstViewer', {
       treeDataProvider: this.treeDataProvider,
       showCollapseAll: true
     });
 
-    vscode.commands.registerCommand('codeQLAstViewer.clear', () => {
+    commands.registerCommand('codeQLAstViewer.clear', () => {
       this.clear();
     });
+
+    ctx.subscriptions.push(window.onDidChangeTextEditorSelection(this.updateTreeSelection, this));
   }
 
   updateRoots(roots: RootAstItem[], db: DatabaseItem, fileName: string) {
     this.treeDataProvider.roots = roots;
     this.treeDataProvider.db = db;
     this.treeDataProvider.refresh();
-    this.treeView.message = `AST for ${fileName}`;
-    this.treeView.reveal(roots[0], { focus: true });
+    this.treeView.message = `AST for ${path.basename(fileName)}`;
+    this.treeView.reveal(roots[0], { focus: false });
+    this.currentFile = fileName;
+  }
+
+  private updateTreeSelection(e: TextEditorSelectionChangeEvent) {
+    function isInside(selectedRange: Range, astRange?: Range): boolean {
+      return !!astRange?.contains(selectedRange);
+    }
+
+    // Recursively iterate all children until we find the node with the smallest
+    // range that contains the selection.
+    // Some nodes do not have a location, but their children might, so must
+    // recurse though location-less AST nodes to see if children are correct.
+    function findBest(selectedRange: Range, items?: RootAstItem[]): RootAstItem | undefined {
```

Comment:
I was concerned about performance but this is pretty snappy on ~1000 LOC. Let's keep an eye on how it performs for much larger files.

Reply:
Yes, I considered adding a debouncing parameter and delaying navigation for that reason, but I didn't notice any issues doing it the simple way.